### PR TITLE
Improve hostname handling in some operations

### PR DESCRIPTION
**Pull request by kwart**
Possible fix for #19165

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterJoinManager.java [hunk 1]
```diff
@@ -683,6 +683,16 @@ private boolean checkIfJoinRequestFromAnExistingMember(JoinMessage joinMessage,
             return true;
         }
 
+        // TODO check if the TcpServerControl.process0(TcpServerConnection, Address, Collection<Address>, MemberHandshake)
+        // could be a better place for the following validation
```
Comment (vbekiaris) [inline hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterJoinManager.java:687]: I agree it's best to catch this pathological condition immediately on member handshake.

**Review by vbekiaris — APPROVED** [review]
looks good.

**Comment by ufukyilmaz** [comment]
for fast fail, the alias checking mechanism should also be applied here:
https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/internal/server/tcp/TcpServerContext.java#L216-L221

File: hazelcast/src/main/java/com/hazelcast/internal/server/Server.java [hunk 2]
```diff
@@ -134,4 +135,8 @@ default int connectionCount() {
      */
     void shutdown();
 
+    default boolean doAddressesMatch(Address expected, Address checked) {
+        ServerConnectionManager conManager = getConnectionManager(EndpointQualifier.MEMBER);
```
Comment (ufukyilmaz) [inline hazelcast/src/main/java/com/hazelcast/internal/server/Server.java:139]: minor: could you also add a comment here about that the conManager cannot be null when endpoint qualifier is `EndpointQualifier.MEMBER`?
AFAIU, we just add a precautionary null check here but we didn't expect this value to be null.

File: hazelcast/src/main/java/com/hazelcast/internal/partition/impl/InternalPartitionServiceImpl.java [hunk 3]
```diff
@@ -1269,12 +1269,14 @@ public boolean isMemberMaster(Address address) {
...
+        Address clusterServiceMaster = clusterService.getMasterAddress();
         if (master == null && clusterService.getSize() == 1) {
-            master = clusterService.getMasterAddress();
+            master = clusterServiceMaster;
         }
 
         // address should be the known master by both PartitionService and ClusterService.
-        return address.equals(master) && address.equals(clusterService.getMasterAddress());
+        return node.server.doAddressesMatch(master, address)
+                && (master.equals(clusterServiceMaster) || node.server.doAddressesMatch(clusterServiceMaster, address));
```
Comment (ufukyilmaz) [inline hazelcast/src/main/java/com/hazelcast/internal/partition/impl/InternalPartitionServiceImpl.java:1279]: Why don't we use `Node#isMaster(Address)` method here? Do we want to reduce the number of `clusterService.getMasterAddress()` calls? 
This newly introduced method has no usage at the moment.

**Review by ufukyilmaz — APPROVED** [review]
I just sent some minor comments above. LGTM in general.

**Comment by kwart** [comment]
run-lab-run

**Comment by kwart** [comment]
run-lab-run

**Comment by kwart** [comment]
Thank you for the reviews, Ufuk & Vassilis. Merged.

**Pull request merged**